library/alloc/src/vec.rs

```diff
@@ -1311,35 +1310,7 @@ impl<T> Vec<T> {
         // the hole, and the vector length is restored to the new length.
         //
         let len = self.len();
-        let start = match range.start_bound() {
-            Included(&n) => n,
-            Excluded(&n) => n + 1,
-            Unbounded => 0,
-        };
-        let end = match range.end_bound() {
-            Included(&n) => n + 1,
-            Excluded(&n) => n,
-            Unbounded => len,
-        };
-
-        #[cold]
-        #[inline(never)]
-        fn start_assert_failed(start: usize, end: usize) -> ! {
-            panic!("start drain index (is {}) should be <= end drain index (is {})", start, end);
-        }
-
-        #[cold]
-        #[inline(never)]
-        fn end_assert_failed(end: usize, len: usize) -> ! {
-            panic!("end drain index (is {}) should be <= len (is {})", end, len);
-        }
-
-        if start > end {
-            start_assert_failed(start, end);
-        }
-        if end > len {
-            end_assert_failed(end, len);
-        }
+        let Range { start, end } = self.check_range(range);
```

Comment:
This call here is the problem: this implicitly creates a slice covering the entire vector, which is a problem because that slice aliases with other pointers that existed before and that should remain valid.

Reply:
I proposed a fix in https://github.com/rust-lang/rust/pull/76662.

Reply:
Interesting. Wouldn't this use the safe `Deref` implementation for `Vec`? Even if not, why doesn't the mutable reference to `self` on this method prevent aliasing?

Reply:
It prevents aliasing from safe clients, but `Vec` explicitly wants to support some unsafe clients that keep pointer into the buffer. That is okay as long as all the operations involved are guaranteed not to reallocate the buffer.

Reply:
[See here](https://github.com/rust-lang/rust/blob/17d32770649fefe8efe6510f7dbff6123f544ba6/library/alloc/tests/vec.rs#L1503) for the testcase that shows how unsafe code can rely on this property.

Reply:
Wow, I had no idea `Vec` makes those guarantees. Thanks for the explanation!

Reply:
Many of these guarantees are [documented here](https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#guarantees). I might have gone overboard in the test and check more than is guaranteed, but then it doesn't seem unlikely that people will take this guarantee as far as they can -- so I added a test basically for every operation that I could imagine preserving the buffer. Someone else double-checking this certainly would not hurt.

Reply:
Indeed, I read that passage but not thoroughly enough to notice these subtleties.

> but then it doesn't seem unlikely that people will take this guarantee as far as they can

IIUC, there is still room for some of those calls to be removed. From the guarantees:

> Bulk insertion methods may reallocate, even when not necessary.

Thus, `append`, `extend`, `extend_from_slice`, and sometimes `splice` shouldn't need to be restricted. Of course, keeping them in the test doesn't hurt. :)

Reply:
Yeah... maybe we should add a comment though that this test does not constitute a guarantee. I'll prepare a PR.